src/zenml/utils/tag_utils.py

```diff
@@ -0,0 +1,636 @@
+    # and this call needs to happen during a step execution. We will use the
+    # step context to fetch the run and attach the tags accordingly.
+    elif all(
+        v is None
+        for v in [
+            run,
+            artifact_version_id,
+            artifact_name,
+            artifact_version,
+            pipeline,
+            run_template,
+        ]
+    ):
+        try:
+            from zenml.steps.step_context import get_step_context
+
+            step_context = get_step_context()
+        except RuntimeError:
+            raise ValueError(
+                f"""
+                You are calling 'add_tags()' outside of a step execution. 
+                If you would like to add tags to a ZenML entity outside 
+                of the step execution, please provide the required 
+                identifiers.\n{add_tags_warning}
+                """
+            )
+
+        # Tag the pipeline run, not the step
+        resource_id = step_context.pipeline_run.id
+        resource_type = TaggableResourceTypes.PIPELINE_RUN
+
+    else:
+        raise ValueError(
+            f"""
+            Unsupported way to call the `add_tags`. Possible combinations "
+            include: \n{add_tags_warning}
+            """
+        )
+
+    # Create tag resources and add tags
+    for tag in tags:
+        try:
+            if isinstance(tag, Tag):
+                tag_model = client.get_tag(tag.name)
+
+                if tag.exclusive and resource_type not in [
+                    TaggableResourceTypes.PIPELINE_RUN,
+                    TaggableResourceTypes.ARTIFACT_VERSION,
+                    TaggableResourceTypes.RUN_TEMPLATE,
+                ]:
+                    raise ValueError(
+                        f"Exclusive tags are only applicable to pipeline runs, "
+                        f"artifact versions and run templates, not {resource_type}."
+                    )
+                if tag.exclusive != tag_model.exclusive:
+                    raise ValueError(
+                        f"The tag `{tag.name}` is an "
+                        f"{'exclusive' if tag_model.exclusive else 'non-exclusive'} "
+                        "tag. Please update it before attaching it to a resource."
+                    )
+                if tag.cascade is not None:
+                    raise ValueError(
+                        "Cascading tags can only be used with the "
+                        "pipeline decorator."
+                    )
+            else:
+                tag_model = client.get_tag(tag)
+
+            if resource_id:
+                client.attach_tag(
+                    tag_name_or_id=tag_model.name,
+                    resources=[
+                        TagResource(id=resource_id, type=resource_type)
+                    ],
+                )
+
+        except KeyError:
+            if isinstance(tag, Tag):
+                tag_model = client.create_tag(name=tag.name)
```

Comment:
Why do we ignore the `ecxlusive` property here? And also all the checks for cascading etc are not handled in this case?

Reply:
Good catch. I added the `Tag` class to the input signature later on and forgot to modify this part of the code. I have added the missing `exclusive` property and the missing checks now.